pkg/auth/providers/azure/clients/ms_graph_client.go

```diff
@@ -31,197 +36,312 @@ const (
-	groups, _, err := c.groupClient.List(context.Background(), odata.Query{Filter: filter})
+func (c AzureMSGraphClient) ListGroups(filter string) ([]v3.Principal, error) {
+	logrus.Debugf("[%s] ListGroups %s", providerLogPrefix, filter)
+	result, err := c.GraphClient.Groups().Get(context.Background(), &msgraphgroups.GroupsRequestBuilderGetRequestConfiguration{
+		QueryParameters: &msgraphgroups.GroupsRequestBuilderGetQueryParameters{
+			Filter: &filter,
+		}})
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("listing groups: %w", err)
 	}
-	var principals []v3.Principal
-	for _, u := range *groups {
-		principal, err := c.groupToPrincipal(u)
-		if err != nil {
-			return nil, err
-		}
-		principals = append(principals, principal)
+
+	pageIterator, err := msgraphcore.NewPageIterator[models.Groupable](
+		result, c.GraphClient.GetAdapter(), models.CreateGroupCollectionResponseFromDiscriminatorValue)
+	if err != nil {
+		return nil, fmt.Errorf("iterating over group list: %w", err)
 	}
-	return principals, err
+
+	var groups []v3.Principal
+	err = pageIterator.Iterate(context.Background(), func(group models.Groupable) bool {
+		groups = append(groups, groupToPrincipal(group))
+		return true
+	})
+
+	return groups, err
+
 }
 
 // ListGroupMemberships takes a user ID and fetches the user's group principals as string IDs from the Microsoft Graph API.
-func (c azureMSGraphClient) ListGroupMemberships(id string) ([]string, error) {
-	groups, _, err := c.userClient.ListGroupMemberships(context.Background(), id, odata.Query{})
+func (c AzureMSGraphClient) ListGroupMemberships(userID string) ([]string, error) {
+	logrus.Debugf("[%s] ListGroupMemberships %s", providerLogPrefix, userID)
+	var groupIDs []string
+	err := c.listGroupMemberships(userID, func(g *models.Group) error {
+		if id := g.GetId(); id != nil {
+			groupIDs = append(groupIDs, *id)
+		}
+		return nil
+	})
 	if err != nil {
 		return nil, err
 	}
-	var names []string
-	for _, g := range *groups {
-		if g.ID != nil && g.DisplayName != nil && g.SecurityEnabled != nil {
-			names = append(names, *g.ID)
-		}
+
+	return groupIDs, nil
+}
+
+func (c AzureMSGraphClient) listGroupMemberships(userID string, f func(*models.Group) error) error {
+	result, err := c.GraphClient.Users().
+		ByUserId(userID).
+		MemberOf().
+		Get(context.Background(), nil)
+	if err != nil {
+		return fmt.Errorf("listing group memberships: %w", err)
 	}
-	return names, nil
+
+	pageIterator, err := msgraphcore.NewPageIterator[models.DirectoryObjectable](
+		result, c.GraphClient.GetAdapter(),
+		models.CreateDirectoryObjectCollectionResponseFromDiscriminatorValue)
+	if err != nil {
+		return fmt.Errorf("iterating over group membership list: %w", err)
+	}
+
+	err = pageIterator.Iterate(context.Background(), func(do models.DirectoryObjectable) bool {
+		group, ok := do.(*models.Group)
+		if !ok {
+			return true
+		}
+		if err := f(group); err != nil {
```

Comment:
We are apparently returning `true` regardless of error or not.
This should allow us to rewrite this and the following lines to
```
_ = f(group)
return true
```

Or is `return true` for errors an error instead ?
Might require a comment to explain what the `return true` means and why issues (`err`, `!ok`) are swallowed.

Reply:
https://pkg.go.dev/github.com/microsoftgraph/msgraph-sdk-go-core#PageIterator.Iterate

`true` means "continue to iterate"

Reply:
The new code looks better.